core/src/main/java/org/elasticsearch/action/fieldstats/TransportFieldStatsAction.java

```diff
@@ -195,26 +193,29 @@ protected FieldStatsShardResponse shardOperation(FieldStatsShardRequest request)
         MapperService mapperService = indexServices.mapperService();
         IndexShard shard = indexServices.getShard(shardId.id());
         try (Engine.Searcher searcher = shard.acquireSearcher("fieldstats")) {
+            // Resolve patterns and deduplicate
+            Set<String> fieldNames = new HashSet<>();
             for (String field : request.getFields()) {
-                Collection<String> matchFields;
-                if (Regex.isSimpleMatchPattern(field)) {
-                    matchFields = mapperService.simpleMatchToIndexNames(field);
-                } else {
-                    matchFields = Collections.singleton(field);
+                fieldNames.addAll(mapperService.simpleMatchToIndexNames(field));
+            }
+            for (String field : fieldNames) {
+                MappedFieldType fieldType = mapperService.fullName(field);
+                if (fieldType == null) {
+                    // ignore missing fields
+                    continue;
                 }
-                for (String matchField : matchFields) {
-                    MappedFieldType fieldType = mapperService.fullName(matchField);
-                    if (fieldType == null) {
-                        // ignore.
-                        continue;
-                    }
-                    FieldStats<?> stats = fieldType.stats(searcher.reader());
+                BytesReference cacheKey = new BytesArray("fieldstats:" + field);
+                BytesReference statsRef = indicesService.cacheShardLevelResult(shard, searcher.getDirectoryReader(), cacheKey, out -> {
```

Comment:
I'll just make it private. If we have a compelling reason to un-private it later we can.